tgym/envs/trading_tick.py

```diff
@@ -0,0 +1,256 @@
+                     '  Price: ' + "%.2f" % self._price + 
+                     '  Tick:' + "%.2f" % self._iteration )
+        self._f.tight_layout()
+        plt.xticks(range(self._iteration)[::5])
+        plt.xlim([max(0, self._iteration - 80.5), self._iteration + 0.5])
+        plt.subplots_adjust(top=0.85)
+        plt.pause(0.01)
+        if savefig:
+            plt.savefig(filename)
+
+    def _get_observation(self):
+        """Concatenate all necessary elements to create the observation.
+
+        Returns:
+            numpy.array: observation array.
+        """
+        return np.concatenate(
+            [prices for prices in self._prices_history[-self._history_length:]] +
+            [
+                np.array([self._price]),
+                np.array(self._position)
+            ]
+        )
+
+    def _add_position(self,action,pr,ls):
+        if all(action == self._actions['hold']): 
+            self._position = self._positions['flat']
+            return 0
+        else:
+            self._total_pnl += 1
+            self._current_pnl += 1
+            self._price = (self._tick_buy if all(action == self._actions['buy']) else self._tick_sell)
+            bt = ('buy' if all(action == self._actions['buy']) else 'sell',self._price, pr, ls,self._total_pnl,self._iteration )
+            self._holding_position.append(bt)
+            self._position = (self._positions['long'] if all(action == self._actions['buy']) else self._positions['short']) 
+            return -self._trading_fee
+    
+    def _close_position(self, position):
+        middle_price = (self._tick_buy +self._tick_sell)/2.0
+        pips = round((middle_price - position[1]) * self._reward_factor, self._round_digits)
+        pr = self._profit_taken
+        st = self._stop_loss
+        if position[0] == 'sell':
+            pips = -pips
+
+        if pips >= pr:
+            #print(position, pips, 'profit_taken',self._tick_buy,self._tick_sell)
+            self._current_pnl -= 1
+            return pr, True
+        elif pips <= st:
+            #print(position, pips,'stop_lost',self._tick_buy,self._tick_sell)
+            self._current_pnl -= 1
+            return st, True
+        else :
+            #print(position,pips, 'nothing',self._tick_buy,self._tick_sell)
+            return 0, False
+
+    def _calculate_position(self):
+        rewards = 0
+        result = []
+        for position in self._holding_position:
+            reward, remove = self._close_position(position)
+            rewards += reward
+            if not remove:
+                result.append(position)
+                
+        self._holding_position = result
+        return rewards
+
+    @staticmethod
+    def random_action_fun():
+        """The default random action for exploration.
+        We hold 80% of the time and buy or sell 10% of the time each.
+
+        Returns:
+            numpy.array: array with a 1 on the action index
```

Comment:
Looks like the end of your file is missing. 
I usually use this to generate a random action (if that ever helps):
```
np.eye(3)[np.random.choice(3, p=[0.95, 0.025, 0.025])]
```
(thx for your interest in this!)

Reply:
https://github.com/cove9988/TradingGym